server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES85BloomFilterPostingsFormat.java

```diff
@@ -513,9 +534,25 @@ static int numBytesForBloomFilter(int bloomFilterSize) {
         return Math.toIntExact((bloomFilterSize + 7L) / 8L);
     }
 
-    static int hashTerm(BytesRef br) {
-        final int hash = murmurhash3_x86_32(br.bytes, br.offset, br.length, 0x9747b28c);
-        return hash & 0x7FFF_FFFF;
+    static final int[] V1_HASH_SEEDS = { 0x9747b28c };
+
+    // Uses CityHash64 to generate a 64-bit hash value, then picks 7 subsets of 31 bits each and returns the values in the
+    // outputs array. This provides us with 7 reasonably independent hashes of the data for the cost of one CityHash64 calculation.
+    static int[] hashTerm(BytesRef br, int[] outputs) {
+        if (outputs.length == 1) {
+            final int hash_v1 = murmurhash3_x86_32(br.bytes, br.offset, br.length, V1_HASH_SEEDS[0]) & 0x7FFF_FFFF;
+            outputs[0] = hash_v1;
+        } else if (outputs.length == 7) {
+            final long hash_v2 = CityHash.cityHash64(br.bytes, br.offset, br.length);
+            outputs[0] = (int) ((hash_v2 & 0x7FFF_FFFF_0000_0000L) >> 32);
+            outputs[1] = (int) ((hash_v2 & 0x07FF_FFFF_F000_0000L) >> (32 - 4));
+            outputs[2] = (int) ((hash_v2 & 0x007F_FFFF_FF00_0000L) >> (32 - 8));
+            outputs[3] = (int) ((hash_v2 & 0x0007_FFFF_FFF0_0000L) >> (32 - 12));
+            outputs[4] = (int) ((hash_v2 & 0x0000_7FFF_FFFF_0000L) >> (32 - 16));
+            outputs[5] = (int) ((hash_v2 & 0x0000_07FF_FFFF_F000L) >> (32 - 20));
+            outputs[6] = (int) ((hash_v2 & 0x0000_007F_FFFF_FF00L) >> (32 - 24));
```

Comment:
I wonder if this is the best way to create 7 hashes from a 64-bits value. Assuming 1.6M docs in a segment so 16M bits in the bloom filter, the bloom filter will use the lower 24 bits of each hash. With this approach the higher 8 bits of the 64-bits hash as well as the lower 8 bits never get used in any of the 7 hashes.

Reply:
This [paper](https://www.eecs.harvard.edu/~michaelm/postscripts/rsa2008.pdf) suggests taking `h1 = (int) (hash_v2 >>> 32)`, `h2 = (int) hash_v2` and computing each of the `i` hashes as `h1 + i * h2`.

Reply:
Good point - the multiplication will mix bits nicely. Given that I have no idea what the maximum size of a segment is, I had made two (likely false, in retrospect) assumptions: (1) segment sizes in excess of 160m documents are possible, which means that more than 24 bits are going to be used, and (2) if the total size of the bloom filter is not a power-of-two, the modular reduction of hash % size will ensure some mixing of the upper bits into the lower bits.

Either way - I think the construction with the multiplication that you are suggesting is superior, and multiplications are plenty fast, so let's do that.